# The dataframe with three columns: PostalCode, Borough, and Neighborhood

## Import Libraries

We will need requests for getting the HTML contents of the website and lxml.html for parsing the relevant fields. Finally, we will store the data on a Pandas Dataframe.

In [1]:
import requests
import lxml.html as lh
import pandas as pd
import numpy as np
from io import StringIO

## Scrape Table Cells

The code below allows us to get Toronto Postal codes data of the HTML table at https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M,

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

For sanity check, ensure that all the rows have the same width. If not, we probably got something more than just the table.

In [3]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

Looks like all our rows have exactly 3 columns. This means all the data collected on tr_elements are from the table.

## Parse Table HeaderParse Table Header

Next, let’s parse the first row as our header.

In [4]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    #print '%d:"%s"'%(i,name)
    print (i,name)
    col.append((name,[]))

1 Postcode
2 Borough
3 Neighbourhood



## Creating Pandas DataFrame

Each header is appended to a tuple along with an empty list.

In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content()
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

Just to be sure, let’s check the length of each column. Ideally, they should all be the same.

In [6]:
[len(C) for (title,C) in col]

[287, 287, 287]

This shows that each of our 3 columns has exactly 287 values.

Now we are ready to create the DataFrame:

In [7]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

## Data Wrangling

In [8]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
       df.shape[0]
    )
)

The dataframe has 11 boroughs and 287 neighborhoods.


In [9]:
#Let's save the postal codes
df.to_csv('First_df.csv')

In [10]:
df.columns

Index(['Postcode', 'Borough', 'Neighbourhood\n'], dtype='object')

In [11]:
#Let's rename Postcode to PostalCode
df.rename(columns={"Postcode": "PostalCode"}, inplace=True)
df.rename(columns={"Neighbourhood\n": "Neighborhood"}, inplace=True)

In [12]:
df.columns

Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

In [13]:
# Let's take the trailing newlines and spaces away from data

df['PostalCode'] = df['PostalCode'].str.strip()
df['Borough'] = df['Borough'].str.strip()
df['Neighborhood'] = df['Neighborhood'].str.strip()

In [14]:
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [15]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

So far so good


In [16]:
#Lets' drop all cells with a borough that is Not assigned and reset index
df.drop(df[df['Borough']  == 'Not assigned'].index, inplace = True) 
df = df.sort_values('PostalCode', ascending=True)
df.reset_index()

index PostalCode      Borough              Neighborhood
0       10        M1B  Scarborough                     Rouge
1       11        M1B  Scarborough                   Malvern
2       28        M1C  Scarborough                Port Union
3       27        M1C  Scarborough                Rouge Hill
4       26        M1C  Scarborough            Highland Creek
5       41        M1E  Scarborough                 Guildwood
6       42        M1E  Scarborough               Morningside
7       43        M1E  Scarborough                 West Hill
8       52        M1G  Scarborough                    Woburn
9       61        M1H  Scarborough                 Cedarbrae
10      75        M1J  Scarborough       Scarborough Village
11      90        M1K  Scarborough      East Birchmount Park
12      91        M1K  Scarborough                   Ionview
13      92        M1K  Scarborough              Kennedy Park
14     107        M1L  Scarborough               Golden Mile
15     108        M1L  Scarborough                  Oakridge
16     106        M1L  Scarborough                  Clairlea
17     122        M1M  Scarborough                Cliffcrest
18     124        M1M  Scarborough  Scarborough Village West
19     123        M1M  Scarborough                 Cliffside
20     140        M1N  Scarborough            Cliffside West
21     139        M1N  Scarborough               Birch Cliff
22     152        M1P  Scarborough           Wexford Heights
23     150        M1P  Scarborough               Dorset Park
24     151        M1P  Scarborough   Scarborough Town Centre
25     163        M1R  Scarborough                  Maryvale
26     164        M1R  Scarborough                   Wexford
27     179        M1S  Scarborough                 Agincourt
28     191        M1T  Scarborough                  Sullivan
29     190        M1T  Scarborough            Clarks Corners
..     ...        ...          ...                       ...
180    285        M8Z    Etobicoke            South of Bloor
181    284        M8Z    Etobicoke     Royal York South West
182      9        M9A    Etobicoke          Islington Avenue
183     21        M9B    Etobicoke                Cloverdale
184     25        M9B    Etobicoke           West Deane Park
185     24        M9B    Etobicoke          Princess Gardens
186     23        M9B    Etobicoke              Martin Grove
187     22        M9B    Etobicoke                 Islington
188     40        M9C    Etobicoke         Old Burnhamthorpe
189     39        M9C    Etobicoke             Markland Wood
190     38        M9C    Etobicoke                  Eringate
191     37        M9C    Etobicoke         Bloordale Gardens
192    121        M9L   North York             Humber Summit
193    138        M9M   North York                 Humberlea
194    137        M9M   North York                     Emery
195    149        M9N         York                    Weston
196    162        M9P    Etobicoke                 Westmount
197    177        M9R    Etobicoke          Richview Gardens
198    175        M9R    Etobicoke         Kingsview Village
199    178        M9R    Etobicoke              St. Phillips
200    176        M9R    Etobicoke      Martin Grove Gardens
201    227        M9V    Etobicoke            Albion Gardens
202    228        M9V    Etobicoke          Beaumond Heights
203    229        M9V    Etobicoke                Humbergate
204    230        M9V    Etobicoke                 Jamestown
205    231        M9V    Etobicoke               Mount Olive
206    233        M9V    Etobicoke             South Steeles
207    234        M9V    Etobicoke               Thistletown
208    232        M9V    Etobicoke               Silverstone
209    244        M9W    Etobicoke                 Northwest

[210 rows x 4 columns]

In [17]:
df.head()

PostalCode      Borough    Neighborhood
10        M1B  Scarborough           Rouge
11        M1B  Scarborough         Malvern
28        M1C  Scarborough      Port Union
27        M1C  Scarborough      Rouge Hill
26        M1C  Scarborough  Highland Creek

In [18]:
df.shape

(210, 3)

In [19]:
#If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

for ind in df.index: 
    if df['Neighborhood'][ind] == 'Not assigned':
        df['Neighborhood'][ind]= df['Borough'][ind]

In [20]:
df

PostalCode      Borough              Neighborhood
10         M1B  Scarborough                     Rouge
11         M1B  Scarborough                   Malvern
28         M1C  Scarborough                Port Union
27         M1C  Scarborough                Rouge Hill
26         M1C  Scarborough            Highland Creek
41         M1E  Scarborough                 Guildwood
42         M1E  Scarborough               Morningside
43         M1E  Scarborough                 West Hill
52         M1G  Scarborough                    Woburn
61         M1H  Scarborough                 Cedarbrae
75         M1J  Scarborough       Scarborough Village
90         M1K  Scarborough      East Birchmount Park
91         M1K  Scarborough                   Ionview
92         M1K  Scarborough              Kennedy Park
107        M1L  Scarborough               Golden Mile
108        M1L  Scarborough                  Oakridge
106        M1L  Scarborough                  Clairlea
122        M1M  Scarborough                Cliffcrest
124        M1M  Scarborough  Scarborough Village West
123        M1M  Scarborough                 Cliffside
140        M1N  Scarborough            Cliffside West
139        M1N  Scarborough               Birch Cliff
152        M1P  Scarborough           Wexford Heights
150        M1P  Scarborough               Dorset Park
151        M1P  Scarborough   Scarborough Town Centre
163        M1R  Scarborough                  Maryvale
164        M1R  Scarborough                   Wexford
179        M1S  Scarborough                 Agincourt
191        M1T  Scarborough                  Sullivan
190        M1T  Scarborough            Clarks Corners
..         ...          ...                       ...
285        M8Z    Etobicoke            South of Bloor
284        M8Z    Etobicoke     Royal York South West
9          M9A    Etobicoke          Islington Avenue
21         M9B    Etobicoke                Cloverdale
25         M9B    Etobicoke           West Deane Park
24         M9B    Etobicoke          Princess Gardens
23         M9B    Etobicoke              Martin Grove
22         M9B    Etobicoke                 Islington
40         M9C    Etobicoke         Old Burnhamthorpe
39         M9C    Etobicoke             Markland Wood
38         M9C    Etobicoke                  Eringate
37         M9C    Etobicoke         Bloordale Gardens
121        M9L   North York             Humber Summit
138        M9M   North York                 Humberlea
137        M9M   North York                     Emery
149        M9N         York                    Weston
162        M9P    Etobicoke                 Westmount
177        M9R    Etobicoke          Richview Gardens
175        M9R    Etobicoke         Kingsview Village
178        M9R    Etobicoke              St. Phillips
176        M9R    Etobicoke      Martin Grove Gardens
227        M9V    Etobicoke            Albion Gardens
228        M9V    Etobicoke          Beaumond Heights
229        M9V    Etobicoke                Humbergate
230        M9V    Etobicoke                 Jamestown
231        M9V    Etobicoke               Mount Olive
233        M9V    Etobicoke             South Steeles
234        M9V    Etobicoke               Thistletown
232        M9V    Etobicoke               Silverstone
244        M9W    Etobicoke                 Northwest

[210 rows x 3 columns]

In [21]:
#How many unique postal codes there are in the duplicates list?
duplicates = df[df.duplicated(['PostalCode'])]
unique_list = []     
unique_list=duplicates.PostalCode.unique()

In [22]:
duplicates

PostalCode      Borough              Neighborhood
11         M1B  Scarborough                   Malvern
27         M1C  Scarborough                Rouge Hill
26         M1C  Scarborough            Highland Creek
42         M1E  Scarborough               Morningside
43         M1E  Scarborough                 West Hill
91         M1K  Scarborough                   Ionview
92         M1K  Scarborough              Kennedy Park
108        M1L  Scarborough                  Oakridge
106        M1L  Scarborough                  Clairlea
124        M1M  Scarborough  Scarborough Village West
123        M1M  Scarborough                 Cliffside
139        M1N  Scarborough               Birch Cliff
150        M1P  Scarborough               Dorset Park
151        M1P  Scarborough   Scarborough Town Centre
164        M1R  Scarborough                   Wexford
190        M1T  Scarborough            Clarks Corners
192        M1T  Scarborough             Tam O'Shanter
206        M1V  Scarborough                  Milliken
205        M1V  Scarborough           L'Amoreaux East
204        M1V  Scarborough           Agincourt North
76         M2J   North York                  Fairview
78         M2J   North York                    Oriole
110        M2L   North York                York Mills
125        M2M   North York               Newtonbrook
30         M3C   North York           Flemingdon Park
65         M3H   North York            Wilson Heights
64         M3H   North York           Downsview North
80         M3J   North York           York University
94         M3K   North York               CFB Toronto
15         M4B    East York             Parkview Hill
..         ...          ...                       ...
254        M8X    Etobicoke              The Kingsway
266        M8Y    Etobicoke          King's Mill Park
265        M8Y    Etobicoke                Humber Bay
270        M8Y    Etobicoke        The Queensway East
271        M8Y    Etobicoke     Royal York South East
272        M8Y    Etobicoke                  Sunnylea
268        M8Y    Etobicoke                 Mimico NE
267        M8Y    Etobicoke  Kingsway Park South East
282        M8Z    Etobicoke                 Mimico NW
283        M8Z    Etobicoke        The Queensway West
285        M8Z    Etobicoke            South of Bloor
284        M8Z    Etobicoke     Royal York South West
25         M9B    Etobicoke           West Deane Park
24         M9B    Etobicoke          Princess Gardens
23         M9B    Etobicoke              Martin Grove
22         M9B    Etobicoke                 Islington
39         M9C    Etobicoke             Markland Wood
38         M9C    Etobicoke                  Eringate
37         M9C    Etobicoke         Bloordale Gardens
137        M9M   North York                     Emery
175        M9R    Etobicoke         Kingsview Village
178        M9R    Etobicoke              St. Phillips
176        M9R    Etobicoke      Martin Grove Gardens
228        M9V    Etobicoke          Beaumond Heights
229        M9V    Etobicoke                Humbergate
230        M9V    Etobicoke                 Jamestown
231        M9V    Etobicoke               Mount Olive
233        M9V    Etobicoke             South Steeles
234        M9V    Etobicoke               Thistletown
232        M9V    Etobicoke               Silverstone

[107 rows x 3 columns]

In [23]:
len(unique_list)

56

In [24]:
gk = df.groupby('PostalCode') 
  
# Let's print the first entries in all the groups formed. 
result=gk.first()
result
for i in result.index:
      print(result)

                     Borough  \
PostalCode                     
M1B              Scarborough   
M1C              Scarborough   
M1E              Scarborough   
M1G              Scarborough   
M1H              Scarborough   
M1J              Scarborough   
M1K              Scarborough   
M1L              Scarborough   
M1M              Scarborough   
M1N              Scarborough   
M1P              Scarborough   
M1R              Scarborough   
M1S              Scarborough   
M1T              Scarborough   
M1V              Scarborough   
M1W              Scarborough   
M1X              Scarborough   
M2H               North York   
M2J               North York   
M2K               North York   
M2L               North York   
M2M               North York   
M2N               North York   
M2P               North York   
M2R               North York   
M3A               North York   
M3B               North York   
M3C               North York   
M3H               North York   
M3J     

In [25]:
#Apply groupby and do the concatenating. More than one neighborhood can exist in one postal code area and they need to be combined into one row with the neighborhoods separated with a comma
df['Neighborhood'] = df[['PostalCode', 'Borough','Neighborhood']].groupby(['PostalCode'])['Neighborhood'].transform(lambda x: ','.join(x))
df[['PostalCode','Borough','Neighborhood']].drop_duplicates()

PostalCode           Borough  \
10         M1B       Scarborough   
28         M1C       Scarborough   
41         M1E       Scarborough   
52         M1G       Scarborough   
61         M1H       Scarborough   
75         M1J       Scarborough   
90         M1K       Scarborough   
107        M1L       Scarborough   
122        M1M       Scarborough   
140        M1N       Scarborough   
152        M1P       Scarborough   
163        M1R       Scarborough   
179        M1S       Scarborough   
191        M1T       Scarborough   
207        M1V       Scarborough   
235        M1W       Scarborough   
245        M1X       Scarborough   
62         M2H        North York   
77         M2J        North York   
93         M2K        North York   
109        M2L        North York   
126        M2M        North York   
141        M2N        North York   
153        M2P        North York   
165        M2R        North York   
2          M3A        North York   
13         M3B        North York   
31         M3C        North York   
63         M3H        North York   
79         M3J        North York   
..         ...               ...   
34         M6C              York   
48         M6E              York   
57         M6G  Downtown Toronto   
71         M6H      West Toronto   
85         M6J      West Toronto   
101        M6K      West Toronto   
116        M6L        North York   
134        M6M              York   
145        M6N              York   
158        M6P      West Toronto   
172        M6R      West Toronto   
185        M6S      West Toronto   
7          M7A  Downtown Toronto   
173        M7R       Mississauga   
264        M7Y      East Toronto   
225        M8V         Etobicoke   
243        M8W         Etobicoke   
256        M8X         Etobicoke   
269        M8Y         Etobicoke   
281        M8Z         Etobicoke   
9          M9A         Etobicoke   
21         M9B         Etobicoke   
40         M9C         Etobicoke   
121        M9L        North York   
138        M9M        North York   
149        M9N              York   
162        M9P         Etobicoke   
177        M9R         Etobicoke   
227        M9V         Etobicoke   
244        M9W         Etobicoke   

                                          Neighborhood  
10                                       Rouge,Malvern  
28                Port Union,Rouge Hill,Highland Creek  
41                     Guildwood,Morningside,West Hill  
52                                              Woburn  
61                                           Cedarbrae  
75                                 Scarborough Village  
90           East Birchmount Park,Ionview,Kennedy Park  
107                      Golden Mile,Oakridge,Clairlea  
122      Cliffcrest,Scarborough Village West,Cliffside  
140                         Cliffside West,Birch Cliff  
152  Wexford Heights,Dorset Park,Scarborough Town C...  
163                                   Maryvale,Wexford  
179                                          Agincourt  
191              Sullivan,Clarks Corners,Tam O'Shanter  
207  Steeles East,Milliken,L'Amoreaux East,Agincour...  
235                                    L'Amoreaux West  
245                                        Upper Rouge  
62                                   Hillcrest Village  
77                          Henry Farm,Fairview,Oriole  
93                                     Bayview Village  
109                            Silver Hills,York Mills  
126                             Willowdale,Newtonbrook  
141                                   Willowdale South  
153                                    York Mills West  
165                                    Willowdale West  
2                                            Parkwoods  
13                                     Don Mills North  
31                     Don Mills South,Flemingdon Park  
63       Bathurst Manor,Wilson Heights,Downsview North  
79                      Northwood Park,York University  
..  

In [26]:
#Sorting the dataframe by postal code and dropping duplicates
df = df.sort_values('PostalCode', ascending=True)
df = df.drop_duplicates(subset='PostalCode', keep='first')
df

PostalCode           Borough  \
10         M1B       Scarborough   
28         M1C       Scarborough   
41         M1E       Scarborough   
52         M1G       Scarborough   
61         M1H       Scarborough   
75         M1J       Scarborough   
92         M1K       Scarborough   
107        M1L       Scarborough   
122        M1M       Scarborough   
140        M1N       Scarborough   
152        M1P       Scarborough   
164        M1R       Scarborough   
179        M1S       Scarborough   
191        M1T       Scarborough   
207        M1V       Scarborough   
235        M1W       Scarborough   
245        M1X       Scarborough   
62         M2H        North York   
78         M2J        North York   
93         M2K        North York   
109        M2L        North York   
126        M2M        North York   
141        M2N        North York   
153        M2P        North York   
165        M2R        North York   
2          M3A        North York   
13         M3B        North York   
30         M3C        North York   
63         M3H        North York   
79         M3J        North York   
..         ...               ...   
34         M6C              York   
48         M6E              York   
57         M6G  Downtown Toronto   
71         M6H      West Toronto   
85         M6J      West Toronto   
101        M6K      West Toronto   
118        M6L        North York   
134        M6M              York   
145        M6N              York   
158        M6P      West Toronto   
172        M6R      West Toronto   
186        M6S      West Toronto   
7          M7A  Downtown Toronto   
173        M7R       Mississauga   
264        M7Y      East Toronto   
225        M8V         Etobicoke   
243        M8W         Etobicoke   
255        M8X         Etobicoke   
269        M8Y         Etobicoke   
284        M8Z         Etobicoke   
9          M9A         Etobicoke   
21         M9B         Etobicoke   
37         M9C         Etobicoke   
121        M9L        North York   
138        M9M        North York   
149        M9N              York   
162        M9P         Etobicoke   
178        M9R         Etobicoke   
233        M9V         Etobicoke   
244        M9W         Etobicoke   

                                          Neighborhood  
10                                       Rouge,Malvern  
28                Port Union,Rouge Hill,Highland Creek  
41                     Guildwood,Morningside,West Hill  
52                                              Woburn  
61                                           Cedarbrae  
75                                 Scarborough Village  
92           East Birchmount Park,Ionview,Kennedy Park  
107                      Golden Mile,Oakridge,Clairlea  
122      Cliffcrest,Scarborough Village West,Cliffside  
140                         Cliffside West,Birch Cliff  
152  Wexford Heights,Dorset Park,Scarborough Town C...  
164                                   Maryvale,Wexford  
179                                          Agincourt  
191              Sullivan,Clarks Corners,Tam O'Shanter  
207  Steeles East,Milliken,L'Amoreaux East,Agincour...  
235                                    L'Amoreaux West  
245                                        Upper Rouge  
62                                   Hillcrest Village  
78                          Henry Farm,Fairview,Oriole  
93                                     Bayview Village  
109                            Silver Hills,York Mills  
126                             Willowdale,Newtonbrook  
141                                   Willowdale South  
153                                    York Mills West  
165                                    Willowdale West  
2                                            Parkwoods  
13                                     Don Mills North  
30                     Don Mills South,Flemingdon Park  
63       Bathurst Manor,Wilson Heights,Downsview North  
79                      Northwood Park,York University  
..  

In [27]:
#Let's verify there is no duplicates
duplicates = df[df.duplicated(['PostalCode'])]


In [28]:
duplicates

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [29]:
#Let's reset index and drop the old index
df = df.reset_index()
del df['index']

In [30]:
#Now the dataframe should be sorted by postal code and reindexed
df

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                        Rouge,Malvern  
1                 Port Union,Rouge Hill,Highland Creek  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7                        Golden Mile,Oakridge,Clairlea  
8        Cliffcrest,Scarborough Village West,Cliffside  
9                           Cliffside West,Birch Cliff  
10   Wexford Heights,Dorset Park,Scarborough Town C...  
11                                    Maryvale,Wexford  
12                                           Agincourt  
13               Sullivan,Clarks Corners,Tam O'Shanter  
14   Steeles East,Milliken,L'Amoreaux East,Agincour...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                          Henry Farm,Fairview,Oriole  
19                                     Bayview Village  
20                             Silver Hills,York Mills  
21                              Willowdale,Newtonbrook  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                     Don Mills South,Flemingdon Park  
28       Bathurst Manor,Wilson Heights,Downsview North  
29                      Northwood Park,York University  
..  

In [31]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
       df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [32]:
#Let's save the postal codes
df.to_csv('TorontoPostalCodes_df.csv')

In [35]:
#Let's see what is the shape of the dataframe now
df.shape

(103, 3)